### PR TITLE
Sentry rageshakes

### PR DESCRIPTION
**Pull request by novocaine**
Fixes https://github.com/vector-im/element-web/issues/11111

Bridge our current rageshake reporting to [Sentry](https://sentry.io) so that equivalent rageshake data can be loaded into a Sentry instance, as well as the current rageshake issues repo. When a rageshake is saved, a Sentry error report is uploaded asynchronously immediately following the regular upload.

Sentry has the following advantages over the current rageshake issues approach:

* It records stack traces, with support for sourcemaps
* It aggregates errors together that have similar stack traces, so you can see which errors are being commonly encountered
* It has a lot of search and filtering functionality for digging through issues
* It has support for workflow stuff like assigning issues to peoples etc

Note that this integration does *not* enable Sentry telemetry whenever an error is encountered - we only send data to Sentry when the user has explicitly opted in to send us debug logs in relation to a problem they have experienced. When the user does that, this the copy we show them:

> Debug logs contain application usage data including your username, the IDs or aliases of the rooms or groups you have visited, which UI elements you last interacted with, and the usernames of other users. They do not contain messages.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Sentry rageshakes ([\#6597](https://github.com/matrix-org/matrix-react-sdk/pull/6597)). Fixes vector-im/element-web#11111. Contributed by [novocaine](https://github.com/novocaine).

Preview: https://611d321f09140a00a3736bbc--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/structures/MatrixChat.tsx [hunk 1]
```diff
@@ -393,6 +394,8 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
         PosthogAnalytics.instance.updatePlatformSuperProperties();
 
         CountlyAnalytics.instance.enable(/* anonymous = */ true);
+
+        initSentry(SdkConfig.get()["sentry"]);
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.tsx:398]: Needs documenting in https://github.com/vector-im/element-web/blob/develop/docs/config.md

File: src/sentry.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
+import * as Sentry from "@sentry/browser";
```
Comment (t3chguy) [inline src/sentry.ts:1]: missing copyright header

**Comment by novocaine** [comment]
Example submission: https://sentry.matrix.org/sentry/riot-web/issues/227300/?query=is%3Aunresolved

File: src/sentry.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            !!(pkCache && await pkCache.getCrossSigningKeyCache("user_signing"))),
+        "secret_storage_ready": String(await client.isSecretStorageReady()),
+        "secret_storage_key_in_account": String(!!(await secretStorage.hasKey())),
+        "session_backup_key_in_secret_storage": String(!!(await client.isKeyBackupKeyStored())),
+        "session_backup_key_cached": String(!!sessionBackupKeyFromCache),
+        "session_backup_key_well_formed": String(sessionBackupKeyFromCache instanceof Uint8Array),
+    };
+}
+
+function getDeviceContext(client: MatrixClient): Record<string, string> {
```
Comment (Palid) [inline src/sentry.ts:101]: Can you type it somehow, so it won't be a `Record<string, string>` but at least a `Record<string, keyof SOMETHING>`? Especially that there's so many keys in here that can or can not exist, adding a proper type here would be very useful.

File: src/components/structures/MatrixChat.tsx (hunk 1 above)
Comment (novocaine) [reply src/components/structures/MatrixChat.tsx:398]: https://github.com/vector-im/element-web/pull/18608

File: src/sentry.ts [hunk 4]
```diff
@@ -21,7 +21,52 @@ import { MatrixClientPeg } from "./MatrixClientPeg";
...
+    storageManager_usageDetails?: string;
+};
+
+type UserContext = {
+    username: string;
+    enabled_labs: string;
+    low_bandwidth: string;
+};
+
+type CryptoContext = {
```
Comment (Palid) [inline src/sentry.ts:39]: Just a heads up, you can instead use `Partial<type>` where you had all the types optional and don't need to repeat `?` everywhere.

**Review by Palid — APPROVED** [review]

**Pull request merged**